crates/interpreter/src/interpreter/shared_memory.rs

```diff
@@ -85,53 +84,44 @@ impl SharedMemory {
     #[cfg(feature = "memory_limit")]
     #[inline]
     pub fn limit_reached(&self, new_size: usize) -> bool {
-        (self.last_checkpoint() + new_size) as u64 > self.memory_limit
+        (self.last_checkpoint + new_size) as u64 > self.memory_limit
     }
 
     /// Prepares the shared memory for a new context.
     #[inline]
-    pub fn new_context_memory(&mut self) {
-        let base_offset = self.last_checkpoint();
-        let new_checkpoint = base_offset + self.current_len;
-
+    pub fn new_context(&mut self) {
+        let new_checkpoint = self.buffer.len();
         self.checkpoints.push(new_checkpoint);
-        self.current_len = 0;
+        self.last_checkpoint = new_checkpoint;
     }
 
     /// Prepares the shared memory for returning to the previous context.
     #[inline]
-    pub fn free_context_memory(&mut self) {
+    pub fn free_context(&mut self) {
         if let Some(old_checkpoint) = self.checkpoints.pop() {
-            let last_checkpoint = self.last_checkpoint();
-            self.current_len = old_checkpoint - last_checkpoint;
+            self.last_checkpoint = self.checkpoints.last().cloned().unwrap_or_default();
+            // safety: buffer len is less than or equal `old_checkpoint`
```

Comment:
```suggestion
            // SAFETY: buffer length is less than or equal `old_checkpoint`
```

SAFETY comment stile is from here: https://std-dev-guide.rust-lang.org/policy/safety-comments.html

Reply:
Thanks for the reference! Soon I'll solve the comments